vllm/core/block/cpu_gpu_block_allocator.py

```diff
@@ -178,6 +176,61 @@ def get_num_free_blocks(self, device: Device) -> int:
         """
         return self._allocators[device].get_num_free_blocks()
 
+    def get_device_related_block_id(self, device: Device,
+                                    absolute_id: int) -> int:
+        """Returns the relative block id on certain device given the absolute 
+        block id.
+
+        Args:
+            device (Device): The device for which to query relative block id.
+                absolute_id (int): The absolute block id for the block in 
+                whole allocator.
+
+        Returns:
+            int: The relative block id on certain device.
+        """
+        return self._allocators[device].get_device_related_block_id(
+            absolute_id)
+
+    def swap(self, blocks: List[Block], source_device: Device,
+             dest_device: Device) -> None:
+        """Execute the swap for the given blocks from source_device
+        on to dest_device, and save the swap mapping.
+
+        Args:
+            blocks: List of blocks to be swapped.
+            source_device (Device): Device to swap the 'blocks' from.
+            dest_device (Device): Device to swap the 'blocks' to.
+        """
+        source_block_ids = [block.block_id for block in blocks]
+        self._allocators[source_device].swap_out(blocks)
+        self._allocators[dest_device].swap_in(blocks)
+        dest_block_ids = [block.block_id for block in blocks]
+        self._swap_mapping = {
+            src: dest
+            for src, dest in zip(source_block_ids, dest_block_ids)
+        }
```

Comment:
Yes, i think this also depends on how we aggregate the multiple swapping results in scheduler layer, since "the same source block id gets mapped to multiple dest blocks" can happen there as well (prefix caching not support in CPU, eg). Let me double check the cache copying kernel and see what we can do there. Will do it tonight.

Reply:
Oh I see -- https://github.com/vllm-project/vllm/blob/6ef09b08f88b675f84b7140238286e5d4c5304c8/vllm/core/scheduler.py#L1075-L1076

Seems this case isn't supported by V1 block manager either. It's OK to stick to the same semantics as today's scheduler (just overwrite in order of observance)

Reply:
So it seems we don't even need `get_and_reset_swaps` and maintain the `self._swap_mapping` per scheduling anymore? As we plan to follow the old semantics.